src/main/scala/is/hail/methods/UpperIndexBounds.scala

```diff
@@ -0,0 +1,65 @@
+package is.hail.methods
+
+import is.hail.HailContext
+import is.hail.annotations.UnsafeIndexedSeq
+import is.hail.expr.types._
+import is.hail.linalg.BlockMatrix
+import is.hail.table.Table
+import is.hail.utils.fatal
+import org.apache.spark.sql.Row
+
+object UpperIndexBounds {
+
+  def groupPositionsByContig(tbl: Table): Array[Array[Int]] = {
+    val expectedSignature = TStruct("contig" -> TString(), "pos" -> TInt32())
+    if (tbl.signature != expectedSignature) {
+      fatal(s"Expected table to have signature $expectedSignature, but found ${ tbl.signature }.")
+    }
+    
+    val groupedPositions = tbl.keyBy("contig").groupByKey("positions").collect()
+      .map(r => r.get(1).asInstanceOf[Vector[Row]].toArray.map(r=>r.get(0).asInstanceOf[Int]))
+
+    groupedPositions.foreach(scala.util.Sorting.quickSort)
+    groupedPositions
+  }
+
+  // positions is non-decreasing, radius is non-negative
+  // for each index i, compute the largest index j such that positions[j]-positions[i] <= radius
+  def computeUpperIndexBounds(positions: Array[Int], radius: Int): Array[Int] = {
+    var contenders = positions.zipWithIndex.tail
```

Comment:
as discussed, try working directly with the `positions` array and a few integer state variables.